contracts/examples/DXswapSwapRelayer.sol

```diff
@@ -0,0 +1,296 @@
+        emit NewOrder(orderIndex);
+    }
+
+    function executeOrder(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(orderIndex < orderCount, 'DXswapRelayer: INVALID_ORDER');
+        require(!order.executed, 'DXswapRelayer: ORDER_EXECUTED');
+        require(oracleCreator.isOracleFinalized(order.oracleId), 'DXswapRelayer: OBSERVATION_RUNNING');
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+
+        address tokenIn = order.tokenIn;
+        uint256 amountOut;
+        amountOut = oracleCreator.consult(
+            order.oracleId,
+            tokenIn == address(0) ? IDXswapRouter(dxSwapRouter).WETH() : tokenIn,
+            order.amountIn
+        );
+
+        uint256 minAmountOut = amountOut.sub(amountOut.mul(order.priceTolerance) / PARTS_PER_MILLION);
+
+        order.executed = true;
+
+        _swap(tokenIn, order.tokenOut, order.amountIn, minAmountOut, order.factory);
+
+        emit ExecutedOrder(orderIndex);
+    }
+
+    // Updates a price oracle and sends a bounty to msg.sender
+    function updateOracle(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(block.timestamp <= order.deadline, 'DXswapRelayer: DEADLINE_REACHED');
+        require(!oracleCreator.isOracleFinalized(order.oracleId), 'DXswapRelayer: OBSERVATION_ENDED');
+        uint256 amountBounty = GAS_ORACLE_UPDATE.mul(tx.gasprice).add(BOUNTY);
+
+        (uint256 reserveA, uint256 reserveB, ) = IDXswapPair(order.oraclePair).getReserves();
+        require(reserveA >= order.minReserveA && reserveB >= order.minReserveB, 'DXswapRelayer: RESERVE_TOO_LOW');
+
+        oracleCreator.update(order.oracleId);
+        if (address(this).balance >= amountBounty) {
+            TransferHelper.safeTransferETH(msg.sender, amountBounty);
+        }
+    }
+
+    function withdrawExpiredOrder(uint256 orderIndex) external {
+        Order storage order = orders[orderIndex];
+        require(msg.sender == owner, 'DXswapRelayer: CALLER_NOT_OWNER');
+        require(block.timestamp > order.deadline, 'DXswapRelayer: DEADLINE_NOT_REACHED');
+        require(order.executed == false, 'DXswapRelayer: ORDER_EXECUTED');
+        address tokenIn = order.tokenIn;
+        uint256 amountIn = order.amountIn;
+        order.executed = true;
+
+        if (tokenIn == address(0)) {
+            TransferHelper.safeTransferETH(owner, amountIn);
+        } else {
+            TransferHelper.safeTransfer(tokenIn, owner, amountIn);
+        }
+        emit WithdrawnExpiredOrder(orderIndex);
+    }
+
+    function _swap(
+        address _tokenIn,
+        address _tokenOut,
+        uint256 _amountIn,
+        uint256 _minAmountOut,
+        address _factory
+    ) internal {
+        uint256[] memory amounts;
+        address[] memory path = new address[](2);
+        address swapRouter;
+
+        if (_tokenIn == address(0)) {
```

Comment:
Preeeetty sure we can leave this to the router and do a check instead. If `tokenIn` is ETH, we can use `swapExactTokensForETH`, `swapExactTokensForTokens` instead. Just a tip though, not a hard requirement.